zml/tokenizer.zig

```diff
@@ -120,16 +122,69 @@ pub const Tokenizer = struct {
         return self.token_lookup.get(str);
     }
 
+    fn toHex(dst: []u8, src: []const u8, comptime case: std.fmt.Case) ![]u8 {
+        return std.fmt.bufPrint(dst, "<0x{s}>", .{switch (case) {
+            .lower => std.fmt.fmtSliceHexLower(src),
+            else => std.fmt.fmtSliceHexUpper(src),
+        }});
+    }
+
+    pub fn preprocess(self: *Tokenizer, allocator: std.mem.Allocator, raw: []const u8) ![]const u8 {
+        var out = std.ArrayList(u8).init(allocator);
+
+        var utf8 = (try std.unicode.Utf8View.init(raw)).iterator();
+        while (utf8.nextCodepointSlice()) |codepoint| {
+            // do not handle spaces (normalizer handles)
+            if (std.mem.eql(u8, codepoint, " ")) {
+                try out.appendSlice(codepoint);
+                continue;
+            }
+            // If found in `Tokenizer.lookup`, append it
+            if (self.lookup(codepoint)) |_| {
+                try out.appendSlice(codepoint);
+                continue;
+            }
+
+            const arena = self.arena_state.allocator();
+
+            // If uppercase hex representation is found, append
+            // (e.g. Huggingface models sometimes converts `\n` to `<0x0A>`).
+            var upper_buffer: [100]u8 = undefined;
+            const upper_hex = try toHex(&upper_buffer, codepoint, .upper);
+            if (self.lookup(upper_hex)) |id| {
+                try self.mapped_tokens.put(arena, codepoint, id);
+                try out.appendSlice(upper_hex);
+                continue;
+            }
+
+            // If lowercase hex representation is found, append
+            // (e.g. Huggingface models sometimes converts `\n` to `<0x0A>`).
+            var lower_buffer: [100]u8 = undefined;
+            const lower_hex = try toHex(&lower_buffer, codepoint, .lower);
+            if (self.lookup(lower_hex)) |id| {
+                try self.mapped_tokens.put(arena, codepoint, id);
+                try out.appendSlice(lower_hex);
+                continue;
+            }
+
+            // Otherwise, append the codepoint as is (will become `<unk>`).
+            try out.appendSlice(codepoint);
+        }
+        return out.toOwnedSlice();
+    }
+
     pub const EncodeOptions = struct {
         /// Should the beginning of sentence '<s>' token be added.
         add_bos: bool = true,
         add_eos: bool = false,
         pad_to: u32 = 0,
     };
 
-    pub fn encode(self: *const Tokenizer, allocator: std.mem.Allocator, raw: []const u8, options: EncodeOptions) ![]u32 {
+    pub fn encode(self: *Tokenizer, allocator: std.mem.Allocator, raw: []const u8, options: EncodeOptions) ![]u32 {
         // log.debug("Tokenizer.encode('{s}')", .{raw});
-        const input = if (self.normalizer) |n| try n.normalize(allocator, raw) else raw;
+        const preprocessed_raw = try self.preprocess(allocator, raw);
```

Comment:
the normalizer should work first, and I'm not sure we need a preprocess at all.

Reply:
So the normalizer does work, but it's not producing the same results as the python normalizer. For [OpenLlama3b](https://huggingface.co/meta-llama/Llama-3.1-8B-Instruct/discussions/35), when testing the snippet on the README, just calling the tokenizer's normalizer on the example prompt, `Q: What is the largest animal?\nA`, yields the following: `['▁Q', ':', '▁What', '▁is', '▁the', '▁largest', '▁animal', '?', '<0x0A>', 'A', ':']`.

This is inconsistent with the behavior of the current Tokenizer implementation.